debezium-core/src/main/java/io/debezium/pipeline/source/spi/ChangeEventSource.java

```diff
@@ -19,12 +21,44 @@ interface ChangeEventSourceContext {
          */
         boolean isRunning();
 
+        /**
+         * Called to indicate that the snapshot has been completed and that streaming should therefore continue.
+         */
         void resumeStreaming() throws InterruptedException;
 
+        /**
+         * Wait for the resumeStreaming function to be called, which indicates that a snapshot is done
+         * and that streaming should resume.
+         */
         void waitSnapshotCompletion() throws InterruptedException;
 
+        /**
+         * Wait for the resumeStreaming function to be called, which indicates that a snapshot is done
+         * and that streaming should resume.
+         *
+         * @param heartbeatCallback A callback function which will be periodically called while waiting for
+         *                          the snapshot to be completed.  Implementations should be kept simple and
+         *                          relatively fast: only do activities like generating activity on the
+         *                          streaming database connection to prevent idle timeouts.  Implementations
+         *                          should also check their own ElapsedTimeStrategy to control how often any
+         *                          heartbeat activities actually occur.
+         */
+        void waitSnapshotCompletion(HeartbeatRunnable heartbeatCallback) throws SQLException, InterruptedException;
+
+        /**
+         * Called by the StreamingChangeEventSource to indicate that the streaming has now been paused, and
+         * that no streaming records are being processed anymore.
+         */
         void streamingPaused();
 
+        /**
+         * Wait for the streamingPaused function to be called.
+         */
         void waitStreamingPaused() throws InterruptedException;
+
+        @FunctionalInterface
+        interface HeartbeatRunnable {
+            void run() throws SQLException;
```

Comment:
Wondering if we can use `RuntimeException` here and the catch the `SQLException` on the implementation and myabe throw a `DebeziumException`. WDYT?

Reply:
Yeah, that looks better.  I double-checked and tested, and it looks like your proposal will not hurt the ability of the connector to restart if connection probing fails in the callback: it's still considered as a retriable exception.  See the latest commit version!